ForestDB + Encryption - SIGSEGV during database deletion

- CBL 1.4-38
- Genymotion - Google Nexus 4 API 16

Test failure - http://uberjenkins.sc.couchbase.com/job/android-listener-client-sync-gateway/163/testReport/junit/testsuites.listener.shared.client_sg/test_replication/test_replication_with_multiple_client_dbs_and_single_sync_gateway_db/

Full adb output - http://uberjenkins.sc.couchbase.com/job/android-listener-client-sync-gateway/163/artifact/results/logs/LiteServAndroid-test_replication_with_multiple_client_dbs_and_single_sync_gateway_db-2017-02-08%2004%3A13%3A22.754080.txt

```
/Sync    (31578): PusherInternal{http://10.17.6.192:4985/db, push, 062d5}: posting to /_revs_diff
D/Sync    (31578): [sendAsyncRequest()] POST => http://10.17.6.192:4985/db/_revs_diff
V/RemoteRequest(31578): com.couchbase.lite.replicator.RemoteRequestRetry@5376c49c: RemoteRequestRetry created, url: http://10.17.6.192:4985/db/_revs_diff
V/Sync    (31578): com.couchbase.lite.replicator.RemoteRequest {POST, http://10.17.6.192:4985/db/_revs_diff}: RemoteRequest created, url: http://10.17.6.192:4985/db/_revs_diff
D/Sync    (31578): PusherInternal{http://10.17.6.192:4985/db, push, 062d5} [fireTrigger()] => RESUME
V/Sync    (31578): com.couchbase.lite.replicator.RemoteRequest {POST, http://10.17.6.192:4985/db/_revs_diff}: RemoteRequest execute() called, url: http://10.17.6.192:4985/db/_revs_diff
V/RemoteRequest(31578): com.couchbase.lite.replicator.RemoteRequest {POST, http://10.17.6.192:4985/db/_revs_diff}: RemoteRequest calling httpClient.execute, url: http://10.17.6.192:4985/db/_revs_diff
I/DEBUG   (  149): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
I/DEBUG   (  149): Build fingerprint: 'generic/vbox86p/vbox86p:4.1.1/JRO03S/eng.genymotion.20160825.115037:userdebug/test-keys'
I/DEBUG   (  149): pid: 31578, tid: 31598, name: Thread: CREATED  >>> com.couchbase.liteservandroid <<<
I/DEBUG   (  149): signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr deadd00d
I/DEBUG   (  149):     eax 0000020c  ebx b6841c98  ecx 00000001  edx 00000000
I/DEBUG   (  149):     esi 97984540  edi 0000020c
I/DEBUG   (  149):     xcs 00000073  xds 0000007b  xes 0000007b  xfs 00000000  xss 0000007b
I/DEBUG   (  149):     eip b679751a  ebp 97984758  esp 97984530  flags 00210246
I/DEBUG   (  149): 
I/DEBUG   (  149): backtrace:
I/DEBUG   (  149):     #00  pc 0005451a  /system/lib/libdvm.so (dvmAbort+170)
I/DEBUG   (  149):     #01  pc 0005b2ae  /system/lib/libdvm.so (addLocalReference(Thread*, Object*)+190)
I/DEBUG   (  149):     #02  pc 00067be4  /system/lib/libdvm.so (GetObjectArrayElement(_JNIEnv*, _jobjectArray*, int)+148)
I/DEBUG   (  149):     #03  pc 000f5f52  /data/data/com.couchbase.liteservandroid/lib/libCouchbaseLiteJavaForestDB.so (Java_com_couchbase_cbforest_DocumentIterator_initEnumerateSomeDocs+162)
I/DEBUG   (  149):     #04  pc 0001aea0  /system/lib/libdvm.so (dvmPlatformInvoke+80)
I/DEBUG   (  149):     #05  pc 000608f9  /system/lib/libdvm.so (dvmCallJNIMethod(unsigned int const*, JValue*, Method const*, Thread*)+489)
I/DEBUG   (  149):     #06  pc 0006bb0d  /system/lib/libdvm.so (dvmResolveNativeMethod(unsigned int const*, JValue*, Method const*, Thread*)+317)
I/DEBUG   (  149):     #07  pc 000c6008  /system/lib/libdvm.so
I/DEBUG   (  149):     #08  pc 00029406  /system/lib/libdvm.so (dvmMterpStd(Thread*)+70)
I/DEBUG   (  149):     #09  pc 000265d0  /system/lib/libdvm.so (dvmInterpret(Thread*, Method const*, JValue*)+160)
I/DEBUG   (  149):     #10  pc 0009111a  /system/lib/libdvm.so (dvmInvokeMethod(Object*, Method const*, ArrayObject*, ArrayObject*, ClassObject*, bool)+938)
I/DEBUG   (  149):     #11  pc 000a20fa  /system/lib/libdvm.so (Dalvik_java_lang_reflect_Method_invokeNative(unsigned int const*, JValue*)+282)
I/DEBUG   (  149):     #12  pc 000c6008  /system/lib/libdvm.so
I/DEBUG   (  149):     #13  pc 00029406  /system/lib/libdvm.so (dvmMterpStd(Thread*)+70)
I/DEBUG   (  149):     #14  pc 000265d0  /system/lib/libdvm.so (dvmInterpret(Thread*, Method const*, JValue*)+160)
I/DEBUG   (  149):     #15  pc 00091779  /system/lib/libdvm.so (dvmCallMethodV(Thread*, Method const*, Object*, bool, JValue*, char*)+601)
I/DEBUG   (  149):     #16  pc 00091886  /system/lib/libdvm.so (dvmCallMethod(Thread*, Method const*, Object*, JValue*, ...)+54)
I/DEBUG   (  149):     #17  pc 00077c33  /system/lib/libdvm.so (interpThreadStart(void*)+755)
I/DEBUG   (  149):     #18  pc 00010870  /system/lib/libc.so (__thread_entry+240)
I/DEBUG   (  149):     #19  pc 00019755  /system/lib/libc.so
I/DEBUG   (  149): 
I/DEBUG   (  149): stack:
I/DEBUG   (  149):          979844f0  00000000  
I/DEBUG   (  149):          979844f4  00000000  
I/DEBUG   (  149):          979844f8  00000000  
I/DEBUG   (  149):          979844fc  00000000  
I/DEBUG   (  149):          97984500  00000000  
I/DEBUG   (  149):          97984504  00000000  
I/DEBUG   (  149):          97984508  00000000  
I/DEBUG   (  149):          9798450c  00000000  
I/DEBUG   (  149):          97984510  00000000  
I/DEBUG   (  149):          97984514  00000000  
I/DEBUG   (  149):          97984518  00000000  
I/DEBUG   (  149):          9798451c  00000000  
I/DEBUG   (  149):          97984520  00000000  
I/DEBUG   (  149):          97984524  00000000  
I/DEBUG   (  149):          97984528  00000000  
I/DEBUG   (  149):          9798452c  46505845  
I/DEBUG   (  149):     #00  97984530  00000000  
I/DEBUG   (  149):          97984534  b6809229  /system/lib/libdvm.so
I/DEBUG   (  149):          97984538  b680dcb7  /system/lib/libdvm.so
I/DEBUG   (  149):          9798453c  00000000  
I/DEBUG   (  149):          97984540  6c756e28  
I/DEBUG   (  149):          97984544  0000296c  
I/DEBUG   (  149):          97984548  00000000  
I/DEBUG   (  149):          9798454c  00000000  
I/DEBUG   (  149):          97984550  00000000  
I/DEBUG   (  149):          97984554  00000000  
I/DEBUG   (  149):          97984558  00000000  
I/DEBUG   (  149):          9798455c  00000000  
I/DEBUG   (  149):          97984560  00000000  
I/DEBUG   (  149):          97984564  00000000  
I/DEBUG   (  149):          97984568  00000000  
I/DEBUG   (  149):          9798456c  00000000  
I/DEBUG   (  149):          ........  ........
I/DEBUG   (  149):     #01  97984760  b86285d8  [heap]
I/DEBUG   (  149):          97984764  00000000  
I/DEBUG   (  149):          97984768  b6810104  /system/lib/libdvm.so
I/DEBUG   (  149):          9798476c  00000200  
I/DEBUG   (  149):          97984770  b86285d8  [heap]
I/DEBUG   (  149):          97984774  00000007  
I/DEBUG   (  149):          97984778  b862863c  [heap]
I/DEBUG   (  149):          9798477c  a6eac0d0  /dev/ashmem/dalvik-heap (deleted)
I/DEBUG   (  149):          97984780  b8839ac0  [heap]
I/DEBUG   (  149):          97984784  b76135e0  /system/lib/libc.so
I/DEBUG   (  149):          97984788  b8784678  [heap]
I/DEBUG   (  149):          9798478c  b6841c98  /system/lib/libdvm.so
I/DEBUG   (  149):          97984790  b86285d8  [heap]
I/DEBUG   (  149):          97984794  000001fe  
I/DEBUG   (  149):          97984798  979847e8  [stack:31598]
I/DEBUG   (  149):          9798479c  b67aabe4  /system/lib/libdvm.so (GetObjectArrayElement(_JNIEnv*, _jobjectArray*, int)+148)
I/DEBUG   (  149):     #02  979847a0  b86285d8  [heap]
I/DEBUG   (  149):          979847a4  dea00005  
I/DEBUG   (  149):          979847a8  979847e8  [stack:31598]
I/DEBUG   (  149):          979847ac  99583678  /data/data/com.couchbase.liteservandroid/lib/libCouchbaseLiteJavaForestDB.so (cbforest::jni::jstringSlice::jstringSlice(_JNIEnv*, _jstring*)+104)
I/DEBUG   (  149):          979847b0  b8839ad8  [heap]
I/DEBUG   (  149):          979847b4  1d2007fd  
I/DEBUG   (  149):          979847b8  979847c7  [stack:31598]
I/DEBUG   (  149):          979847bc  b7611f4c  /system/lib/libc.so
I/DEBUG   (  149):          979847c0  b75a64a9  /system/lib/libc.so (malloc+9)
I/DEBUG   (  149):          979847c4  0170bcb4  
I/DEBUG   (  149):          979847c8  00000000  
I/DEBUG   (  149):          979847cc  99603727  /data/data/com.couchbase.liteservandroid/lib/libCouchbaseLiteJavaForestDB.so (operator new(unsigned int)+39)
I/DEBUG   (  149):          979847d0  00000010  
I/DEBUG   (  149):          979847d4  a6e85558  /dev/ashmem/dalvik-heap (deleted)
I/DEBUG   (  149):          979847d8  97984828  [stack:31598]
I/DEBUG   (  149):          979847dc  9970bcb4  /data/data/com.couchbase.liteservandroid/lib/libCouchbaseLiteJavaForestDB.so
I/DEBUG   (  149):          ........  ........
I/DEBUG   (  149): 
I/DEBUG   (  149): memory map around fault addr deadd00d:
I/DEBUG   (  149):     bfeeb000-bff0c000 [stack]
I/DEBUG   (  149):     (no map for address)
I/DEBUG   (  149):     (no map above)
W/InputDispatcher(  320): channel '537e81d4 com.couchbase.liteservandroid/com.couchbase.liteservandroid.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
E/InputDispatcher(  320): channel '537e81d4 com.couchbase.liteservandroid/com.couchbase.liteservandroid.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
W/InputDispatcher(  320): Attempted to unregister already unregistered input channel '537e81d4 com.couchbase.liteservandroid/com.couchbase.liteservandroid.MainActivity (server)'
I/qtaguid (  320): Failed write_ctrl(s 0 10048) res=-1 errno=1
```

Note to Self:

This test fails also with another reason. Failed to store data because of the concurrent issue. This induces the document count mismatch.

```
02-08 23:01:31.481 9803-9828/com.couchbase.liteservandroid E/CBForest: Forestdb Handle is being used by another thread (-39)
02-08 23:01:31.485 9803-9828/com.couchbase.liteservandroid E/Database: ForestDB Error: forceInsert()
com.couchbase.cbforest.ForestException: Forestdb Handle is being used by another thread
   at com.couchbase.cbforest.Document.init(Native Method)
   at com.couchbase.cbforest.Document.<init>(Document.java:97)
   at com.couchbase.cbforest.Database.getDocument(Database.java:46)
   at com.couchbase.lite.store.ForestDBStore$4.run(ForestDBStore.java:1036)
   at com.couchbase.lite.store.ForestDBStore.inTransaction(ForestDBStore.java:1515)
   at com.couchbase.lite.store.ForestDBStore.forceInsert(ForestDBStore.java:1030)
   at com.couchbase.lite.Database.forceInsert(Database.java:2075)
   at com.couchbase.lite.replicator.PullerInternal$9.run(PullerInternal.java:595)
   at com.couchbase.lite.store.ForestDBStore.runInTransaction(ForestDBStore.java:306)
   at com.couchbase.lite.Database.runInTransaction(Database.java:646)
   at com.couchbase.lite.replicator.PullerInternal.insertDownloads(PullerInternal.java:577)
   at com.couchbase.lite.replicator.PullerInternal$2.process(PullerInternal.java:131)
   at com.couchbase.lite.support.Batcher.processNow(Batcher.java:403)
   at com.couchbase.lite.support.Batcher.access$300(Batcher.java:29)
   at com.couchbase.lite.support.Batcher$2.run(Batcher.java:337)
   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
   at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
   at java.util.concurrent.FutureTask.run(FutureTask.java:137)
   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
   at java.lang.Thread.run(Thread.java:856)
02-08 23:01:31.489 9803-9828/com.couchbase.liteservandroid W/Sync: com.couchbase.lite.replicator.PullerInternal$9@5378cafc: failed to write {ls_db2_446 #1-39d06a6cf4eae404ca2710e433776bb2 @460}: status=-1
02-08 23:01:31.489 9803-9828/com.couchbase.liteservandroid W/Sync: PullerInternal{http://192.168.33.11:4985/db, pull, 793d8}: Progress: set error = com.couchbase.lite.replicator.RemoteRequestResponseException: {code=-1}
02-08 23:01:31.489 9803-9828/com.couchbase.liteservandroid E/CBForest: Forestdb Handle is being used by another thread (-39)
02-08 23:01:31.505 9803-9828/com.couchbase.liteservandroid E/Database: ForestDB Error: forceInsert()
com.couchbase.cbforest.ForestException: Forestdb Handle is being used by another thread
   at com.couchbase.cbforest.Document.init(Native Method)
   at com.couchbase.cbforest.Document.<init>(Document.java:97)
   at com.couchbase.cbforest.Database.getDocument(Database.java:46)
   at com.couchbase.lite.store.ForestDBStore$4.run(ForestDBStore.java:1036)
   at com.couchbase.lite.store.ForestDBStore.inTransaction(ForestDBStore.java:1515)
   at com.couchbase.lite.store.ForestDBStore.forceInsert(ForestDBStore.java:1030)
   at com.couchbase.lite.Database.forceInsert(Database.java:2075)
   at com.couchbase.lite.replicator.PullerInternal$9.run(PullerInternal.java:595)
   at com.couchbase.lite.store.ForestDBStore.runInTransaction(ForestDBStore.java:306)
   at com.couchbase.lite.Database.runInTransaction(Database.java:646)
   at com.couchbase.lite.replicator.PullerInternal.insertDownloads(PullerInternal.java:577)
   at com.couchbase.lite.replicator.PullerInternal$2.process(PullerInternal.java:131)
   at com.couchbase.lite.support.Batcher.processNow(Batcher.java:403)
   at com.couchbase.lite.support.Batcher.access$300(Batcher.java:29)
   at com.couchbase.lite.support.Batcher$2.run(Batcher.java:337)
   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
   at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
   at java.util.concurrent.FutureTask.run(FutureTask.java:137)
   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
   at java.lang.Thread.run(Thread.java:856)
02-08 23:01:31.505 9803-9828/com.couchbase.liteservandroid W/Sync: com.couchbase.lite.replicator.PullerInternal$9@5378cafc: failed to write {ls_db2_447 #1-2388b0607db3b12c27bdb43a2c48befd @461}: status=-1
02-08 23:01:31.505 9803-9828/com.couchbase.liteservandroid W/Sync: PullerInternal{http://192.168.33.11:4985/db, pull, 793d8}: Progress: set error = com.couchbase.lite.replicator.RemoteRequestResponseException: {code=-1}
02-08 23:01:31.521 9803-9821/com.couchbase.liteservandroid E/CBForest: Forestdb Handle is being used by another thread (-39)
02-08 23:01:31.521 9803-9821/com.couchbase.liteservandroid E/Database: Error in getAllDocs()
com.couchbase.cbforest.ForestException: Forestdb Handle is being used by another thread
   at com.couchbase.cbforest.DocumentIterator.next(Native Method)
   at com.couchbase.cbforest.DocumentIterator.next(DocumentIterator.java:40)
   at com.couchbase.cbforest.DocumentIterator.nextDocument(DocumentIterator.java:58)
   at com.couchbase.lite.store.ForestDBStore.readFromIterator(ForestDBStore.java:723)
   at com.couchbase.lite.store.ForestDBStore.getAllDocs(ForestDBStore.java:661)
   at com.couchbase.lite.Database.getAllDocs(Database.java:1664)
   at com.couchbase.lite.router.Router.do_POST_Document_all_docs(Router.java:1219)
   at java.lang.reflect.Method.invokeNative(Native Method)
   at java.lang.reflect.Method.invoke(Method.java:511)
   at com.couchbase.lite.router.Router.start(Router.java:557)
   at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
   at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
   at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
   at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
   at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
   at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
   at java.lang.Thread.run(Thread.java:856)
02-08 23:01:31.525 9803-9821/com.couchbase.liteservandroid W/Router: Router unable to route request to do_POST_Document_all_docs
java.lang.reflect.InvocationTargetException
 at java.lang.reflect.Method.invokeNative(Native Method)
 at java.lang.reflect.Method.invoke(Method.java:511)
 at com.couchbase.lite.router.Router.start(Router.java:557)
 at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
 at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
 at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
 at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
 at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
 at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
 at java.lang.Thread.run(Thread.java:856)
Caused by: java.lang.NullPointerException
 at com.couchbase.lite.router.Router.convertCBLQueryRowsToMaps(Router.java:1165)
 at com.couchbase.lite.router.Router.do_POST_Document_all_docs(Router.java:1220)
 at java.lang.reflect.Method.invokeNative(Native Method) 
 at java.lang.reflect.Method.invoke(Method.java:511) 
 at com.couchbase.lite.router.Router.start(Router.java:557) 
 at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141) 
 at javax.servlet.http.HttpServlet.service(HttpServlet.java:802) 
 at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408) 
 at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330) 
 at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110) 
 at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238) 
 at java.lang.Thread.run(Thread.java:856) 
```

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -1215,12 +1215,11 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         List<Object> keys = (List<Object>) body.get("keys");
         options.setKeys(keys);
 
-        Map<String, Object> result = null;
-        result = db.getAllDocs(options);
-        convertCBLQueryRowsToMaps(result);
-
+        Map<String, Object> result = db.getAllDocs(options);
         if (result == null) {
             return new Status(Status.INTERNAL_SERVER_ERROR);
+        } else {
+            convertCBLQueryRowsToMaps(result);
         }
         connection.setResponseBody(new Body(result));
         return new Status(Status.OK);

